File: question/leetcode/leetcode0253.Meeting-Rooms-II/leetcode0253.py
import heapq
from typing import List


class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0

        # 1. 先按会议开始时间排序
        intervals.sort(key=lambda x: x[0])

        # 2. 初始化一个最小堆，放第一个会议的结束时间
        heap = [intervals[0][1]]  # heap中存的是end时间

        # 3. 遍历剩下的会议
        for start, end in intervals[1:]:
            # 如果最早结束的会议 <= 当前会议的开始时间，复用会议室（弹出堆顶）
            if heap[0] <= start:
                heapq.heappop(heap)
            # 无论复用还是新开，都加入当前会议的结束时间
            heapq.heappush(heap, end)

        # 4. 最终堆大小 = 所需会议室数量
        return len(heap)

File: question/leetcode/leetcode0253.Meeting-Rooms-II/leetcode0253.ts
function minMeetingRooms(intervals: number[][]): number {
	if (intervals.length === 0) {
		return 0;
	}

	//按开始时间排序
	intervals.sort((a, b) => a[0] - b[0]);

	//模拟最小堆,会议结束时间
	const heap: number[] = []

	for (const [start, end] of intervals) {
		//如果当前会议开始时间 >= 最小结束时间，可以复用会议室
		if (heap.length > 0 && start >= heap[0]) {
			//移除最早结束的会议
			heap.shift();
		}
		//加入当前会议的结束时间
		heap.push(end);

		//维持headp为最小堆结构
		heap.sort((a, b) => a - b);

	}

	return heap.length;
};